src/arch/x86/lib/cpu_performance/cpu_info.c

```diff
@@ -7,11 +7,20 @@
  */
 
 #include <stdint.h>
-
+#include <string.h>
+#include <assert.h>
 #include <hal/cpu_info.h>
 
 struct cpu_info current_cpu;
 
+void set_feature_val(struct cpu_info *info, const char *name, unsigned int val) {
+	int current_feature = info->feature_count;
+	assert((current_feature < CURRENT_ARCH_NUM_FEATURES) && "Exceeded Number of Features!");
+	strcpy(info->feature[current_feature].name, name);
+	info->feature[current_feature].val = val;
+	info->feature_count += 1;
+}
```

Comment:
This function is arch-independent, so it can be moved out of this (x86/lib/cpu_info) module.

Reply:
I tried moving it in `hal/cpu_info.h` but I was getting build errors.

Reply:
Why does `x86/lib/cpu_performance/cpu_counter.c` have `#include <hal/cpu_info.h>` ? This was causing double compilation and hence a build failure. Should I remove it?

Reply:
And we don't need the x86 `cpu_info.h` file anymore.

Reply:
Hm.. It should not lead to compilation error; header guard should protect header file from double-include
```
#ifndef HAL_CPU_INFO_H_
#define HAL_CPU_INFO_H_
...
#endif
```

Reply:
Yeah I thought so, but as soon as I removed the include in `cpu_counter.c` the build was successful. And in the other case was indeed recompilation of the function.

Reply:
Ah, you tried to move the whole function.

C preprocessor works straightforward: when you do `#include <file.h>`, it simply substitutes content of the file.

So imagine you have files `f1.c, f2.c, f3.h`,
```
f3:
int some_func() { return 0; }
```
```
f2:
#include <f3.h>

int other_func() { return some_func(); }
```

```
f1:
#include <f3.h>

int some_other_func() { return some_func(); }
```

After preprocessor does its stuff you get:
```
f2:
int some_func() { return 0; }

int other_func() { return some_func(); }
```

```
f1:
int some_func() { return 0; }

int some_other_func() { return some_func(); }
```

After that if you try to link `f1.o` and `f2.o`, linker doesn't know which `some_func()` you actually want to call (non-static functions are visible to all other c-files).

Reply:
As you can see, this problem raises when you define (not just declare) functions in header files.

So there are two solutions:
1) Declare function as `static`, so it will be unique for every C file.
2) Define it in some other C file and replace definition with declaration in the header file.

Reply:
Alright, in which directory should I place the new C file?

Reply:
Well, I guess it's a good idea to create `src/lib/cpu_info` and create a module for this library.

That's pretty narrow scope for a library; may be it will be merged with something else in the future; but for now let's do it this way :)

Reply:
Hey @Deryugin, just committed these changes in a new fixup commit :)